### PR TITLE
generate kubelet auth cert on upgrades

### LINKED ISSUE #42199: GCE Upgrade loses api server certs
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Bug

**What happened**:

Upgrade hangs. API server won't start because it doesn't have valid certs.

**What you expected to happen**:

Upgrade succeeds

**How to reproduce it** (as minimally and precisely as possible):

```console
> go run hack/e2e.go --check_version_skew=false --v=t --up --test --test_args="--ginkgo.focus=\[Feature:MasterUpgrade\] --upgrade-target=ci/latest --upgrade-image=gci"
```

**Anything else we need to know**:

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-latest-upgrade-master/315 is an example. Log from kube-apiserver.log:

```
I0227 22:30:14.951728       5 gce.go:332] Using existing Token Source &oauth2.reuseTokenSource{new:google.computeSource{account:""}, mu:sync.Mutex{state:0, sema:0x0}, t:(*oauth2.Token)(nil)}
unable to load server certificate: tls: failed to find any PEM data in certificate input
```

I reproduced this locally and checked the certs:

```console
krousey@e2e-test-krousey-master /etc/srv/kubernetes/pki $ ls -alF             
total 40
drwxr-xr-x 2 root root 4096 Feb 27 21:53 ./
drwxr-xr-x 5 root root 4096 Feb 27 22:05 ../
-rw-r--r-- 1 root root 4451 Feb 27 22:05 apiserver-client.crt
-rw-r--r-- 1 root root 1708 Feb 27 22:05 apiserver-client.key
-rw-r--r-- 1 root root    0 Feb 27 22:05 apiserver.crt
-rw-r--r-- 1 root root    0 Feb 27 22:05 apiserver.key
-rw-r--r-- 1 root root 1233 Feb 27 22:05 ca-certificates.crt
-rw-r--r-- 1 root root 1233 Feb 27 22:05 ca.crt
-rw-r--r-- 1 root root  227 Feb 27 22:05 ca.key
-rw-r--r-- 1 root root 1233 Feb 27 22:05 kubelet.crt
-rw-r--r-- 1 root root 1679 Feb 27 22:05 kubelet.key
-rw-r--r-- 1 root root    0 Feb 27 22:05 serviceaccount.crt
-rw-r--r-- 1 root root    0 Feb 27 22:05 serviceaccount.key
```

apiserver.{crt,key} are empty.

### PR DESCRIPTION
**Pull request by mikedanese**
cc @cjcullen @liggitt 

fixes #42199

### CODE REVIEW

File: cluster/gce/upgrade.sh [hunk 1]
```diff
@@ -101,6 +102,48 @@ function upgrade-master() {
...
+function backfile-kubeletauth-certs() {
+  if [[ ! -z "${KUBEAPISERVER_CERT_BASE64:-}" && ! -z "${KUBEAPISERVER_CERT_BASE64:-}" ]]; then
+    return 0
+  fi
+
+  mkdir -p "${KUBE_TEMP}/pki"
+  echo "${CA_KEY_BASE64}" | base64 -d > "${KUBE_TEMP}/pki/ca.key"
+  echo "${CA_CERT_BASE64}" | base64 -d > "${KUBE_TEMP}/pki/ca.crt"
+  (cd "${KUBE_TEMP}/pki"
+    download-cfssl
```
Comment (liggitt) [inline cluster/gce/upgrade.sh]: Is this a copy paste from somewhere else? I had trouble finding exactly this elsewhere and expected it to match existing cluster setup scripts

Comment (mikedanese) [inline cluster/gce/upgrade.sh]: This is not a copy paste. The existing scripts split creation between easy-rsa and cfssl for some reason and require creating *everything* in one swoop incluing the CA. We probably want to just move to cfssl for everything and consolidate all this (going to work on that in a followup). This usage seems fairly straightforward though.

**Comment by mikedanese** [comment]
@k8s-bot kops aws e2e test this
@k8s-bot gci gce e2e test this
@k8s-bot gce etcd3 e2e test this

**Comment by skriss** [comment]
@k8s-bot gce etcd3 e2e test this

File: cluster/gce/upgrade.sh [hunk 2]
```diff
@@ -101,6 +102,48 @@ function upgrade-master() {
...
 
+# TODO(mikedanese): delete when we don't support < 1.6
+function backfile-kubeletauth-certs() {
+  if [[ ! -z "${KUBEAPISERVER_CERT_BASE64:-}" && ! -z "${KUBEAPISERVER_CERT_BASE64:-}" ]]; then
+    return 0
+  fi
+
+  mkdir -p "${KUBE_TEMP}/pki"
+  echo "${CA_KEY_BASE64}" | base64 -d > "${KUBE_TEMP}/pki/ca.key"
+  echo "${CA_CERT_BASE64}" | base64 -d > "${KUBE_TEMP}/pki/ca.crt"
```
Comment (liggitt) [inline cluster/gce/upgrade.sh]: Careful about the permissions of key files. Not sure what we do elsewhere

File: cluster/gce/upgrade.sh [hunk 3]
```diff
@@ -101,6 +102,48 @@ function upgrade-master() {
...
+        -profile=client \
+        - \
+      | "${KUBE_TEMP}/cfssl/cfssljson" -bare kube-apiserver
+{
+  "CN": "kube-apiserver"
+}
+EOF
+  )
+  KUBEAPISERVER_CERT_BASE64=$(cat "${KUBE_TEMP}/pki/kube-apiserver.pem" | base64 | tr -d '\r\n')
+  KUBEAPISERVER_KEY_BASE64=$(cat "${KUBE_TEMP}/pki/kube-apiserver-key.pem" | base64 | tr -d '\r\n')
```
Comment (liggitt) [inline cluster/gce/upgrade.sh]: Same about key permissions

File: cluster/gce/upgrade.sh [hunk 4]
```diff
@@ -101,6 +102,48 @@ function upgrade-master() {
...
+    cat <<EOF \
+      | "${KUBE_TEMP}/cfssl/cfssl" gencert \
+        -ca=ca.crt \
+        -ca-key=ca.key \
+        -config=ca-config.json \
+        -profile=client \
+        - \
+      | "${KUBE_TEMP}/cfssl/cfssljson" -bare kube-apiserver
+{
+  "CN": "kube-apiserver"
```
Comment (liggitt) [inline cluster/gce/upgrade.sh]: Comment that this connects to the rolebinding allowing the API server to talk to the kubelet

**Comment by liggitt** [comment]
A few nits, looks pretty reasonable. I assume it's been manually tested since merge queue jobs don't hit this.

**Comment by skriss** [comment]
@mikedanese @liggitt can we get this approved tonight? beta.2 is being held for it at the moment so we can get GCE upgrade testing working

File: cluster/gce/upgrade.sh (hunk 2 above)
Comment (mikedanese) [inline cluster/gce/upgrade.sh]: These files are generated on the work station running kube-up.sh, and ${KUBE_TEMP} is deleted in the trap EXIT handler. We don't do any umasking elsewhere.

https://github.com/kubernetes/kubernetes/blob/master/cluster/common.sh#L987-L996

File: cluster/gce/upgrade.sh (hunk 3 above)
Comment (mikedanese) [inline cluster/gce/upgrade.sh]: See above. Following precedent.

File: cluster/gce/upgrade.sh (hunk 4 above)
Comment (mikedanese) [inline cluster/gce/upgrade.sh]: Done.

**Comment by skriss** [comment]
@k8s-bot kops aws e2e test this

**Pull request merged**